### PR TITLE
Adding ingress resource to federation apiserver

### PR DESCRIPTION
**Pull request by nikhiljindal**
Fixes https://github.com/kubernetes/kubernetes/issues/29342

cc @kubernetes/sig-cluster-federation

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/30112)

### CODE REVIEW

File: federation/pkg/federation-controller/namespace/namespacecontroller.go [hunk 1]
```diff
@@ -0,0 +1,917 @@
+/*
```
Comment (madhusudancs) [inline federation/pkg/federation-controller/namespace/namespacecontroller.go]: This code doesn't belong to this PR, I think?

File: test/e2e/federation-ingress.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
+/*
+Copyright 2015 The Kubernetes Authors.
```
Comment (madhusudancs) [inline test/e2e/federation-ingress.go]: 2016?

File: test/e2e/federation-ingress.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
+/*
```
Comment (madhusudancs) [inline test/e2e/federation-ingress.go]: `federated-ingress.go`, the verb form, meaning the Ingress itself is federated, as opposed to an Ingress that is just part of a federation.

File: test/e2e/federation-ingress.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/pkg/apis/extensions/v1beta1"
+	"k8s.io/kubernetes/test/e2e/framework"
+
+	. "github.com/onsi/gomega"
+)
+
+const (
+	FederationIngressName = "federation-ingress"
```
Comment (madhusudancs) [inline test/e2e/federation-ingress.go]: Same comment as above. `FederatedIngressName = "federated-ingress"`

File: test/e2e/federation-ingress.go [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+)
+
+// Create/delete ingress api objects
+var _ = framework.KubeDescribe("Federation ingresses [Feature:Federation]", func() {
+	f := framework.NewDefaultFederatedFramework("federation-ingress")
```
Comment (madhusudancs) [inline test/e2e/federation-ingress.go]: Reuse the const defined above.

File: test/e2e/federation-ingress.go [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+			framework.ExpectNoError(err, "Error deleting ingress %q in namespace %q", ingress.Name, ingress.Namespace)
+			By(fmt.Sprintf("Deletion of ingress %q in namespace %q succeeded.", ingress.Name, f.Namespace.Name))
+		})
+
+	})
+})
+
+func createIngressOrFail(clientset *federation_release_1_4.Clientset, namespace string) *v1beta1.Ingress {
+	if clientset == nil || len(namespace) == 0 {
+		Fail(fmt.Sprintf("Internal error: invalid parameters passed to deleteIngressOrFail: clientset: %v, namespace: %v", clientset, namespace))
```
Comment (madhusudancs) [inline test/e2e/federation-ingress.go]: s/deleteIngressOrFail/createIngressOrFail/

File: test/e2e/federation-ingress.go [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+	}
+	By(fmt.Sprintf("Creating federated ingress %q in namespace %q", FederationIngressName, namespace))
+
+	ingress := &v1beta1.Ingress{
+		ObjectMeta: v1.ObjectMeta{
+			Name: FederationIngressName,
+		},
+	}
+
+	By(fmt.Sprintf("Trying to create ingress %q in namespace %q", ingress.Name, namespace))
```
Comment (madhusudancs) [inline test/e2e/federation-ingress.go]: Isn't this just a duplicate of the previous `By()` line?

**Comment by madhusudancs** [comment]
@nikhiljindal looks good mostly. A few minor nits. Feel free to self-apply the label after fixing them.

File: federation/pkg/federation-controller/namespace/namespacecontroller.go (hunk 1 above)
Comment (nikhiljindal) [inline federation/pkg/federation-controller/namespace/namespacecontroller.go]: oops, sorry :)
removed

File: test/e2e/federation-ingress.go (hunk 2 above)
Comment (nikhiljindal) [inline test/e2e/federation-ingress.go]: fixed

File: test/e2e/federation-ingress.go (hunk 6 above)
Comment (nikhiljindal) [inline test/e2e/federation-ingress.go]: fixed

File: test/e2e/federation-ingress.go (hunk 7 above)
Comment (nikhiljindal) [inline test/e2e/federation-ingress.go]: removed

File: test/e2e/federation-ingress.go (hunk 5 above)
Comment (nikhiljindal) [inline test/e2e/federation-ingress.go]: The const happens to have the same value, but it is different. Leaving this as is.

File: test/e2e/federation-ingress.go (hunk 3 above)
Comment (nikhiljindal) [inline test/e2e/federation-ingress.go]: This is a test for ingress resource in federation control plane?
Leaving this as is

**Comment by nikhiljindal** [comment]
Thanks @madhusudancs for the review.
Have updated the code as per comments.
PTAL

File: test/e2e/federation-ingress.go (hunk 3 above)
Comment (madhusudancs) [inline test/e2e/federation-ingress.go]: but it is for a `federated-ingress`.

Comment (madhusudancs) [inline test/e2e/federation-ingress.go]: For reference, see the title of this doc here - https://github.com/kubernetes/kubernetes/pull/29793

File: test/e2e/federation-ingress.go (hunk 4 above)
Comment (madhusudancs) [inline test/e2e/federation-ingress.go]: Same comment as above.

**Comment by madhusudancs** [comment]
@nikhiljindal Just a couple more nits (more of a follow up).

**Comment by nikhiljindal** [comment]
The only remaining comment was to name this "federation-ingress" or "federated-ingress". As discussed offline, "federation-ingress" is fine.
Adding LGTM.
Thanks @madhusudancs
